cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,556 @@
+    pOp->u.publish.pReceivedData = malloc_can_fail( length );
+
+  } else {
+    // allocate operation member of the union
+
+    // assumption here is checked below in valid_*
+    __CPROVER_assume( VALID_CBMC_SIZE( pOp->u.operation.packetSize ) );
+    pOp->u.operation.pMqttPacket =
+      malloc_can_fail( pOp->u.operation.packetSize );
+
+  }
+
+  return pOp;
+}
+
+// TODO: valid_ should have same signature as allocate_,
+// should check that operation points to the connection
+
+bool valid_IotMqttOperation( IotMqttOperation_t pOp )
+{
+  if (pOp == NULL) return 0;
+
+  // check that publish member of the union is valid (pOp->incomingPublish)
+
+  bool valid_publishinfo =
+    valid_IotMqttPublishInfo( &(pOp->u.publish.publishInfo) );
+
+  bool valid_publish_member =
+    IMPLIES( pOp->incomingPublish, valid_publishinfo );
+
+  // check that operation member of the union is valid (!pOp->incomingPublish)
+
+  bool valid_packet =
+    VALID_STRING( pOp->u.operation.pMqttPacket, pOp->u.operation.packetSize ) &&
+    VALID_CBMC_SIZE( pOp->u.operation.packetSize );
+  bool valid_pingreq_packet =
+    IMPLIES( pOp->u.operation.type == IOT_MQTT_PINGREQ,
+	     IFF( pOp->u.operation.periodic.ping.keepAliveMs == 0,
+		  pOp->u.operation.packetSize == 0 ) );
+  bool valid_other_packet =
+    IMPLIES( pOp->u.operation.type != IOT_MQTT_PINGREQ,
+	     pOp->u.operation.packetSize > 0 );
+  bool waitable =
+    (pOp->u.operation.flags & IOT_MQTT_FLAG_WAITABLE) == IOT_MQTT_FLAG_WAITABLE;
+  bool valid_jobReference =
+    IMPLIES(  waitable, pOp->u.operation.jobReference == 2 ) &&
+    IMPLIES( !waitable, pOp->u.operation.jobReference == 1 );
+
+  bool valid_operation_member =
+    IMPLIES(!pOp->incomingPublish,
+	    valid_packet &&
+	    valid_pingreq_packet &&
+	    valid_other_packet &&
+	    valid_jobReference);
+
+  return
+    // assume valid connection
+    valid_publish_member &&
+    valid_operation_member;
+}
+
+/****************************************************************
+ * IotMqttConnection
+ ****************************************************************/
+
+IotMqttConnection_t allocate_IotMqttConnection(IotMqttConnection_t pConn)
+{
+  if ( pConn == NULL ) pConn = malloc_can_fail( sizeof( *pConn ) );
+  if ( pConn == NULL ) return NULL;
+
+  pConn->pNetworkConnection = allocate_IotNetworkConnection();
+  pConn->pNetworkInterface = allocate_IotNetworkInterface();
+  allocate_IotMqttOperation( &(pConn->pingreq), pConn );
+  return pConn;
+}
+
+void ensure_IotMqttConnection_has_lists(IotMqttConnection_t pConn)
+{
+  // TODO: add code to make lists nondet nontrivial
+  // Consider using allocate_IotMqttSubscriptionList
+  IotListDouble_Create(&pConn->pendingProcessing);
+  IotListDouble_Create(&pConn->pendingResponse);
+  IotListDouble_Create(&pConn->subscriptionList);
+  return pConn;
+}
+
+bool valid_IotMqttConnection(IotMqttConnection_t pConn)
+{
+  if ( pConn == NULL ) return 0;
+
+  bool valid_references = pConn->references >= 1;
```

Comment:
Add a comment to explain magic number "1".

Reply:
Fixed.  I had pushed to the wrong repository.

Reply:
Still shows as 1 for me

Reply:
Added a comment explaining the member.